Задача BJS2-66564 выполнена

gulnaz21 opened the pull request.

Commit:
Задача BJS2-66575 выполнена

Commit:
Внесены правки и добавлены тесты

gulnaz21 committed:
Задача BJS2-66564 выполнена

Zhltsk-V commented on src/main/java/faang/school/accountservice/model/Balance.java:
```diff
@@ -0,0 +1,75 @@
+    @Column(name = "version", nullable = false)
+    private Long version;
+
+    public void authorize(BigDecimal amount) {
```
Методы authorize, clear, deposit, cancelAuthorization содержат бизнес-логику. Рассмотри вариант переноса этих методов в BalanceService или BalanceHelper.

gulnaz21 replied:
здесь логика простая и относится только к состоянию объекта, поэтому вынесла ее в сущность, чтобы не загромождать код в сервисе геттерами и сеттерами

RinatBeybutov replied:
Это хорошее решение. Для себя можно погуглить, что такое DDD. Там много таких нюансов

Zhltsk-V commented on src/main/java/faang/school/accountservice/model/Balance.java:
```diff
@@ -0,0 +1,75 @@
+@NoArgsConstructor
+@Getter
+@Setter
+@Slf4j
```
Аннотация @Slf4j в Balance нигде не используется.

Zhltsk-V commented on src/main/java/faang/school/accountservice/service/balance/BalanceService.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @param accountId идентификатор счета
+     */
+    public void createBalanceForAccount(Long accountId) {
```
Думаю стоит пометить метод createBalanceForAccount аннотацией @Transactional.

gulnaz21 replied:
Я не повесила здесь аннотацию `@transactional`, так как этот метод вызывается только в методе openAccount класса AccountService, который уже помечен данной аннотацией, так что по сути он все равно не создаст новую транзакцию

Zhltsk-V commented on src/main/java/faang/school/accountservice/service/balance/BalanceService.java:
```diff
@@ -0,0 +1,111 @@
+     * @param accountId идентификатор счета
+     * @return доступный баланс
+     */
+    public BigDecimal getAvailableBalance(Long accountId) {
```
Думаю лучше возвращать BalanceViewDto

Zhltsk-V commented on src/main/java/faang/school/accountservice/controller/BalanceController.java:
```diff
@@ -0,0 +1,71 @@
+            summary = "Get available balance",
+            description = "Returns current available balance (actual balance)")
+    @GetMapping("/available-balance")
+    public BigDecimal getAvailableBalance(@NotNull @PathVariable Long accountId) {
```
Можно добавить аннотация @Positive к Long accountId чтобы избежать некорректных значений.

Zhltsk-V commented on src/main/java/faang/school/accountservice/service/balance/BalanceHelper.java:
```diff
@@ -0,0 +1,46 @@
+                });
+    }
+
+    @Transactional
```
Можно указать изоляцию @Transactional(isolation = Isolation.SERIALIZABLE)

gulnaz21 replied:
А для чего? Я думаю, это будет избыточно, здесь ведь и так проверяется версия

RinatBeybutov replied:
изоляция уровня serializable не применяется в промышленной разработке из-за того, что все транзакции выстраиваются в очередь и отсутствует параллельность

Zhltsk-V commented on src/main/java/faang/school/accountservice/repository/BalanceRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.util.Optional;
+
+@Repository
```
Эта аннотация необязательна для наследников JpaRepository

gulnaz21 committed:
fix:
   - исправлены недочеты

RinatBeybutov commented on src/main/java/faang/school/accountservice/repository/AccountRepository.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
лишняя аннотация, есть родитель JpaRepository

RinatBeybutov commented on src/main/java/faang/school/accountservice/service/account/AccountHelper.java:
```diff
@@ -0,0 +1,84 @@
+     * @return учетная запись
+     * @throws AccountNotFoundException если учетная запись не найдена
+     */
+    public Account getAccountById(Long accountId) {
```
не понимаю смысла оставлять этот метод в классе Helper. Думаю, лучше перенести в AccountService

RinatBeybutov commented on src/main/java/faang/school/accountservice/service/account/AccountHelper.java:
```diff
@@ -0,0 +1,84 @@
+     * @throws IllegalStateException если не удалось сгенерировать уникальный номер после 10 попыток
+     */
+    public String generateAccountNumber() {
+        int attempt = 0;
```
давай попробуем порефакторить код:
```
for(int attempt =0; attempt < 10; attempt++) {
	
	int length = ThreadLocalRandom.current().nextInt(12, 21);

	StringBuilder sb = new StringBuilder();
	for (int i = 0; i < length; i++) {
		sb.append(ThreadLocalRandom.current().nextInt(i == 0 ? 1 : 0, 10));
	}
	number = sb.toString();
	
	if(!accountRepository.existsByAccountNumber(number)) {
		return number;
	}
}

throw new IllegalStateException("Не удалось сгенерировать уникальный номер счёта");
```

Если есть идеи, как еще улучшить, то предлагай. Можно еще число 10 вынести в константу класса

RinatBeybutov commented on src/main/java/faang/school/accountservice/service/account/AccountHelper.java:
```diff
@@ -0,0 +1,84 @@
+     * @return сохраненная учетная запись
+     * @throws AccountOperationConflictException если возникает конфликт при сохранении учетной записи
+     */
+    public Account saveAccount(Account account, String errorMessage) {
```
почему этот метод в хелпере, а не в сервисе?
передавать текст ошибки сюда странно кмк

RinatBeybutov commented on src/main/java/faang/school/accountservice/service/account/AccountService.java:
```diff
@@ -0,0 +1,146 @@
+        account.setAccountStatus(AccountStatus.ACTIVE);
+        account.setVersion(0);
+
+        Account savedAccount = accountHelper.saveAccount(account, "Failed to save the account");
```
здесь можно переиспользовать:
account = accountHelper.saveAccount(account, "Failed to save the account");

RinatBeybutov commented on src/main/java/faang/school/accountservice/service/account/AccountService.java:
```diff
@@ -0,0 +1,146 @@
+    public AccountViewDto unblockAccount(Long accountId) {
+        Account account = accountHelper.getAccountById(accountId);
+        accountValidator.validateUnblock(account);
+        return updateAccountStatus(account, AccountStatus.ACTIVE, Instant.now());
```
при разблокировке устанавливается после даты закрытия? что-то странно выглядит

RinatBeybutov commented on src/main/java/faang/school/accountservice/model/Account.java:
```diff
@@ -0,0 +1,80 @@
+    private Instant updatedAt;
+
+    @Column(name = "closed_at")
+    private Instant closedAt;
```
вот тут не хватает java docs. При разблокировке аккаунта нужно ли это поле выставлять?

RinatBeybutov commented on src/main/java/faang/school/accountservice/validation/AccountValidator.java:
```diff
@@ -0,0 +1,98 @@
+     */
+    public void validateBlock(Account account) {
+        validateNotClosed(account);
+        if (account.getAccountStatus() == AccountStatus.BLOCKED) {
```
здесь можно переиспользовать метод validateStatus(account, AccountStatus.BLOCKED)

gulnaz21 replied:
таска с аккаунтом не моя, у меня работа с балансом была

gulnaz21 committed:
Merge branch 'murloc-master-stream9' into feature/BJS2-66564-working-with-balances

# Conflicts:
#	src/main/java/faang/school/accountservice/dto/account/AccountViewDto.java
#	src/main/java/faang/school/accountservice/mapper/AccountMapper.java
#	src/main/java/faang/school/accountservice/model/Account.java
#	src/main/java/faang/school/accountservice/repository/AccountRepository.java
#	src/main/java/faang/school/accountservice/service/account/AccountHelper.java
#	src/main/java/faang/school/accountservice/service/account/AccountService.java
#	src/main/java/faang/school/accountservice/validation/AccountValidator.java
#	src/main/resources/db/changelog/changeset/V001__account-service_account.sql
#	src/test/java/faang/school/accountservice/controller/AccountControllerTest.java
#	src/test/java/faang/school/accountservice/service/account/AccountHelperTest.java
#	src/test/java/faang/school/accountservice/service/account/AccountServiceTest.java
#	src/test/java/faang/school/accountservice/validation/AccountValidatorTest.java

gulnaz21 committed:
fix:
   - решены конфликты
   - переписан метод для генерации счета
   - добавлена JavaDoc
   - исправлена работа с closedAt в методе updateAccountStatus
   - исправлены тесты
   - методы для получения сущностей Account и Balance перенесены в соответствующие сервисы

RinatBeybutov commented on src/main/java/faang/school/accountservice/dto/account/AccountViewDto.java:
```diff
@@ -29,5 +28,5 @@ public class AccountViewDto {
     private Instant closedAt;
     private Integer version;
-    private BigDecimal balance;
+    private Long balance;
```
а дробная часть не нужна?

RinatBeybutov commented on src/main/java/faang/school/accountservice/service/account/AccountHelper.java:
```diff
@@ -93,4 +67,17 @@ public Account saveAccount(Account account) {
         log.debug("Attempting to save account {}", account.getId());
```
Не очень понимаю смысл повторного сохранения с ошибкой. Оптимистичная ошибка возникает в том случае, если в базе уже поменяли значение в то время, как мы что-то меняем в нашем методе. От повторного сохранения у нас не изменится сущность, которую мы хотим сохранить.

RinatBeybutov commented on src/test/java/faang/school/accountservice/controller/BalanceControllerTest.java:
```diff
@@ -0,0 +1,91 @@
+    @MockBean
+    private UserContext userContext;
+
+    Long accountId = 1L;
```
private ?

gulnaz21 committed:
fix:
   - исправлены недочеты

gulnaz21 committed:
для решения конфликтов

gulnaz21 committed:
Merge branch 'murloc-master-stream9' into feature/BJS2-66564-working-with-balances

# Conflicts:
#	src/main/resources/db/changelog/db.changelog-master.yaml

gulnaz21 committed:
fix:
